shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,200 @@
+  successInvocationSelector?: string,
+  totalFailureInvocation?: SbpInvocation
+}
+
+type PersistentActionStatus = {
+  cancelled: boolean,
+  failedAttemptsSoFar: number,
+  lastError: string,
+  nextRetry: string,
+  pending: boolean
+}
+
+const defaultOptions: PersistentActionOptions = {
+  maxAttempts: Number.POSITIVE_INFINITY,
+  retrySeconds: 30
+}
+const tag = '[chelonia.persistentActions]'
+
+class PersistentAction {
+  invocation: SbpInvocation
+  options: PersistentActionOptions
+  status: PersistentActionStatus
+  timer: Object
+
+  constructor (invocation: SbpInvocation, options: PersistentActionOptions = {}) {
+    this.invocation = invocation
+    this.options = { ...defaultOptions, ...options }
+    this.status = {
+      cancelled: false,
+      failedAttemptsSoFar: 0,
+      lastError: '',
+      nextRetry: '',
+      pending: false
+    }
+  }
+
+  // Do not call if the action is pending or cancelled!
+  async attempt (): Promise<void> {
+    if (await this.trySBP(this.options.skipCondition)) {
+      this.cancel()
+      return
+    }
+    try {
+      this.status.pending = true
+      const result = await sbp(...this.invocation)
+      this.handleSuccess(result)
+    } catch (error) {
+      this.handleError(error)
+    }
+  }
+
+  cancel (): void {
+    this.timer && clearTimeout(this.timer)
+    this.status.cancelled = true
+    this.status.nextRetry = ''
+  }
+
+  async handleError (error: Error): Promise<void> {
+    const { options, status } = this
+    // Update relevant status fields before calling any optional selector.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    status.nextRetry = anyAttemptLeft && !status.cancelled
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
+    // Perform any optional SBP invocation.
+    await this.trySBP(options.errorInvocation)
+    !anyAttemptLeft && await this.trySBP(options.totalFailureInvocation)
+    // Schedule a retry if appropriate.
+    if (status.nextRetry) {
+      // Note: there should be no older active timeout to clear.
+      this.timer = setTimeout(() => this.attempt(), this.options.retrySeconds * 1e3)
+    }
+    status.pending = false
+  }
+
+  async handleSuccess (result: any): Promise<void> {
+    const { status } = this
+    status.lastError = ''
+    status.nextRetry = ''
+    status.pending = false
+    this.options.successInvocationSelector &&
+      await this.trySBP([this.options.successInvocationSelector, result])
+  }
+
+  trySBP (invocation: SbpInvocation | void): any {
+    try {
+      return invocation ? sbp(...invocation) : undefined
+    } catch (error) {
+      console.error(tag, error.message)
+    }
+  }
+}
+
+// SBP API
+
+sbp('sbp/selectors/register', {
+  'chelonia.persistentActions/_init' (): void {
+    sbp('okTurtles.events/on', LOGIN, (function () {
```

Comment:
`LOGIN` is special to Group Income, it cannot be used in Chelonia.

Reply:
In that case maybe I can make this module define a public `init` selector, which will be called from within Group Income on login.

Reply:
Most of this stuff can be left in `_init`. E.g. `this.actionsByID = Object.create(null)` etc. are all things that can be done inside of `/_init`.

The only thing that seems to need to be configured at runtime is the `databaseKey`, so similarly to `chelonia` itself you could add a `chelonia.persistentActions/configure` selector where that can be set.